### DISCUSSION TITLE
Dynamic theme with tailwind.config variables in Vue 3

### DISCUSSION THREAD

**Discussion by Drennargh** [tailwindlabs/tailwindcss#12562] [Help] [answered]
Hello everyone,

--------------
Edit: The solution to my problem seems to be this: https://www.reddit.com/r/vuejs/comments/tfrqb9/how_to_style_the_components_root_element_vue3sfc/
Always open for better suggestions.
---------------

I love Tailwind, but now I'm struggling with some more advanced usages and I can't find this issue. Only some partially similar cases.

I'm extending the colors with a primary and secondary color class, each with light, base and dark shade. The colors are defined by variables with a default value retrieved from the  const colors = require("tailwindcss/colors");.

```
// tailwind.config.js

const colors = require("tailwindcss/colors");
....
  theme: {
    extend: {
      colors: {
        "primary-color": {
          light: "var(--colors-primary-light," + colors.gray["100"] + ")",
          base: "var(--colors-primary-base," + colors.gray["200"] + ")",
          dark: "var(--colors-primary-dark," + colors.gray["300"] + ")",
        },
        "secondary-color": {
          light: "var(--colors-secondary-light," + colors.slate["800"] + ")",
          base: "var(--colors-secondary-base," + colors.slate["900"] + ")",
          dark: "var(--colors-secondary-dark," + colors.slate["950"] + ")",
        },

````

This works! When I use the classes like bg-primary-color-light, I get the corresponding color.

I am using Vue.js, so I should be able to define the css variables in the component, right?

The currentTheme I get from a Pinia store via axios call.

```
const currentTheme= computed(() => {
  return themeStore.currentTheme;
});

watch(currentTheme, getColor);

function getColor() {
  if (currentTheme.value.colors.primary.base) {
    color.value = currentTheme.value.colors.primary.base.hex;
  } else {
    //
  }
}
....
<style lang="scss">

:root {
  --colors-primary-light: 100 115 179; //I am using different inputs to see if something's working. Unfortunately not.
  --colors-primary-base: "#ac5529";
  --colors-primary-dark: v-bind("color");
  --colors-secondary-light: v-bind("color");
  --colors-secondary-base: v-bind("color");
  --colors-secondary-dark: v-bind("color");
}
````
This doesn't seem to work.

````
.test-class {
  background-color: v-bind("color");
}
````
This does work.

What's wrong with the variables? Since the second example works, v-bind("colors") is fine.

**Comment by wongjn** [comment]
I'd assume it's to do with the `:root` selector not being compatible with `<style scoped>`. The `scoped` attribute would mean the CSS gets compiled to something like:

```css
[data-v-some-hash]:root {
  /* … */
}
```
Which may not be applying to any element in the DOM (this component would need to be rendering the `<html>` root element), thus the CSS variable values aren't set.

Whereas the working example would compile to:

```css
.test-class[data-v-some-hash] {
  /* … */
}
```
Which *could* apply to a DOM element.

**Reply by Drennargh** [reply]
Thanks for the quick reply wongjn! I forgot to edit that out, because I did test both with and without scoped, but the result remains the same. I edited the "scoped" out of the original post to avoid confusion.

**Reply by wongjn** [reply]
The CSS variable names in your Tailwind configuration file do not seem to match the CSS variable definitions. For example, you have `--colors-primary-color-light` in the `<style>` block but you have `--colors-primary-light` in the Tailwind configuration.

**Reply by Drennargh** [reply]
Sorry, that was a bit of spaghetti as well from trying different combinations. I have correct it in the code. It doesn't work, even with the correct names.

**Reply by wongjn** [reply, marked as answer]
The CSS variables that Vue produce for the `v-bind` in CSS are still component-scoped, whereas `:root` would be an ancestral element and thus would not have the CSS variables from `v-bind` set:

```css
:root {
  --colors-primary-dark: var(--some-hash)
}
```
```html
<!-- --some-hash is currently not defined at this level -->
<html>
  <body>
    <!-- --some-hash is only defined here where the component renders -->
    <div style="--some-hash:your-color">
```

**Reply by Drennargh** [reply]
Thank you wongjn! I have now defined it in the App.vue as

````

<template>
  <div class="theme">
....
  </div>
</template>

and the style:
.theme {
  --colors-primary-light: v-bind("color");
  --colors-primary-base: v-bind("color");
  --colors-primary-dark: v-bind("color");
  --colors-secondary-light: v-bind("color");
  --colors-secondary-base: v-bind("color");
  --colors-secondary-dark: v-bind("color");

  --test-color: v-bind("color");
}
````
This way it shows correctly in all the components. I think it's fine like this.
